File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260921-211557-5789/analysis/own-tests-project/src/main/java/simshotels/Hotel.java
package simshotels;

import java.util.ArrayList;
import java.util.List;
import java.util.function.ToIntFunction;

public class Hotel {

    private final List<Floor> floors;

    // instance creation (class side)

    public static Hotel withFloors(List<Floor> aFloorsCollection) {
        Hotel.assertHaveFloors(aFloorsCollection);

        return new Hotel(aFloorsCollection);
    }

    // initialization

    private Hotel(List<Floor> aFloorsCollection) {
        floors = new ArrayList<>(aFloorsCollection);
    }

    // testing

    public boolean isEmpty() {
        return floors.stream().allMatch(Floor::isAvailable);
    }

    // accounting

    public int totalLosses() {
        return totalOfFloors(Floor::totalLosses);
    }

    public int totalProfits() {
        return totalOfFloors(Floor::totalProfits);
    }

    // totals

    public int totalRooms() {
        return totalOfFloors(Floor::totalRooms);
    }

    public int totalRoomsAvailable() {
        return totalOfFloors(Floor::totalRoomsAvailable);
    }

    public int totalRoomsOccupied() {
        return totalOfFloors(Floor::totalRoomsOccupied);
    }

    public int totalRoomsReserved() {
        return totalOfFloors(Floor::totalRoomsReserved);
    }

    private int totalOfFloors(ToIntFunction<Floor> aTotalOfFloor) {
        return floors.stream().mapToInt(aTotalOfFloor).sum();
    }

    // guests

    public void receiveAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorAt(aFloorNumber).receiveAtRoom(aGuestType, aRoomNumber);
    }

    public void receiveWithReservationAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorAt(aFloorNumber).receiveWithReservationAtRoom(aGuestType, aRoomNumber);
    }

    public void reserveRoomAtFloor(int aRoomNumber, int aFloorNumber) {
        floorAt(aFloorNumber).reserveRoom(aRoomNumber);
    }

    // accessing

    private Floor floorAt(int aFloorNumber) {
        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }

        return floors.get(aFloorNumber - 1);
    }

    // assertions (class side)

    public static void assertHaveFloors(List<Floor> aFloorsCollection) {
        if (aFloorsCollection.isEmpty()) {
            Hotel.signalNoFloors();
        }
    }

    // error description (class side)

    public static String floorNumberDoesNotExistErrorDescription() {
        return "Floor number does not exist";
    }

    public static String noFloorsErrorDescription() {
        return "Cannot have a Hotel without floors";
    }

    public static String roomNumberDoesNotExistErrorDescription() {
        return Floor.roomNumberDoesNotExistErrorDescription();
    }

    // exceptions (class side)

    public static void signalFloorNumberDoesNotExist() {
        throw new RuntimeException(Hotel.floorNumberDoesNotExistErrorDescription());
    }

    public static void signalNoFloors() {
        throw new RuntimeException(Hotel.noFloorsErrorDescription());
    }
}

File: experiments/000-PipelineChecks/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-202040-52205/output/project/src/simshotels/hotel.py
class Hotel:

    # initialization

    def __init__(self):
        self._floors = []

    # accounting

    def total_losses(self):
        return sum(floor.total_losses() for floor in self._floors)

    def total_profits(self):
        return sum(floor.total_profits() for floor in self._floors)

    # setters

    def set_floors(self, a_floors_collection):
        Hotel.assert_have_floors(a_floors_collection)

        self._floors = a_floors_collection

        return self

    # totals

    def total_rooms(self):
        return sum(floor.total_rooms() for floor in self._floors)

    def total_rooms_available(self):
        return sum(floor.total_rooms_available() for floor in self._floors)

    def total_rooms_occupied(self):
        return sum(floor.total_rooms_occupied() for floor in self._floors)

    def total_rooms_reserved(self):
        return sum(floor.total_rooms_reserved() for floor in self._floors)

    # testing

    def is_empty(self):
        return all(floor.is_available() for floor in self._floors)

    # guests

    def receive_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        floor = self._get_floor(a_floor_number)
        floor.receive_at_room(a_guest_type, a_room_number)

    def receive_with_reservation_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        floor = self._get_floor(a_floor_number)
        floor.receive_with_reservation_at_room(a_guest_type, a_room_number)

    def reserve_room_at_floor(self, a_room_number, a_floor_number):
        floor = self._get_floor(a_floor_number)
        floor.reserve_room(a_room_number)

    # private

    def _get_floor(self, a_floor_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()
        return self._floors[a_floor_number - 1]

    # assertions (class side)

    @staticmethod
    def assert_have_floors(a_floors_collection):
        if len(a_floors_collection) == 0:
            Hotel.signal_no_floors()

    # error description (class side)

    @staticmethod
    def floor_number_does_not_exist_error_description():
        return "Floor number does not exist"

    @staticmethod
    def no_floors_error_description():
        return "Cannot have a Hotel without floors"

    @staticmethod
    def room_number_does_not_exist_error_description():
        return "Room number does not exist"

    # exceptions (class side)

    @staticmethod
    def signal_floor_number_does_not_exist():
        raise RuntimeError(Hotel.floor_number_does_not_exist_error_description())

    @staticmethod
    def signal_no_floors():
        raise RuntimeError(Hotel.no_floors_error_description())

    @staticmethod
    def signal_room_number_does_not_exist():
        raise RuntimeError(Hotel.room_number_does_not_exist_error_description())
